### PR TITLE
feature/did_resolver

### PR DESCRIPTION
**Pull request by burdettadam**
Signed-off-by: Adam Burdett <burdettadam@gmail.com>

### CODE REVIEW

**Review by dbluhm — COMMENTED** [review]
I think we should move the did registry and the did resolver from `core` into `resolver`. What do you think?

File: aries_cloudagent/resolver/diddoc.py [hunk 1]
```diff
@@ -93,9 +91,11 @@ def _valid_filter(service):
             *(sorted(old, key=itemgetter("priority"))),
         ]
 
-    def dereference(self, did_url: str):
+    def dereference(self, did_url: Union[str, DIDUrl]):
         """Dereference values contained in this DID Document."""
-        parsed = DIDUrl.parse(did_url)
+        if isinstance(did_url, str):
+            parsed = DIDUrl.parse(did_url)
+
```
Comment (dbluhm) [inline aries_cloudagent/resolver/diddoc.py:98]: Needs an else with `parsed = did_url`

File: aries_cloudagent/core/tests/test_did_resolver.py [hunk 2]
```diff
@@ -51,50 +53,108 @@
...
 
 
-@pytest.mark.parametrize("did", TEST_DIDS)
-def test_resolve(did):
-    did = resolver.resolve(did)
-    assert isinstance(did, ResolvedDIDDoc)
+#@pytest.mark.parametrize("did", TEST_DIDS)
+#def test_resolve(did):
+#    did_doc = await resolver.resolve(did)
+#    assert did_doc
```
Comment (dbluhm) [inline aries_cloudagent/core/tests/test_did_resolver.py:160]: This test file could use some cleaning up

File: aries_cloudagent/core/tests/test_did_resolver.py [hunk 3]
```diff
@@ -51,50 +53,108 @@
...
+def test_dereference_external(did):
+    with async_mock.patch.object(
+            resolver, "resolve",
+            async_mock.MagicMock(return_value="resolved did doc")
+        ) as mock_resolver:
+        result = mock_resolver.dereference_external(did)
+        assert result
 
 
+@pytest.mark.parametrize("doc",DOC)
```
Comment (dbluhm) [inline aries_cloudagent/core/tests/test_did_resolver.py:152]: Parametrizing a single value doesn't make sense. Just use `DOC` directly.

File: aries_cloudagent/core/tests/test_did_resolver.py [hunk 4]
```diff
@@ -51,50 +53,108 @@
...
 @pytest.mark.parametrize("did, method", zip(TEST_DIDS, TEST_DID_METHODS))
 def test_match_did_to_resolver(did, method):
-    base_resolver = resolver._match_did_to_resolver(did)
+    did = DID(did)
+    base_resolver = next(resolver._match_did_to_resolver(did))
     assert base_resolver.supports(method)
 
 
 def test_match_did_to_resolver_native_priority():
+    # TODO: implement this
```
Comment (dbluhm) [inline aries_cloudagent/core/tests/test_did_resolver.py:139]: Using a pytest skip to mark tests that need more work is probably preferable to a todo comment

File: aries_cloudagent/core/tests/test_did_resolver.py [hunk 5]
```diff
@@ -51,50 +53,108 @@
...
     test_resolver_1,
     test_resolver_2,
     test_resolver_3,
     test_resolver_4,
 ]
-for resolver in resolvers:
+for resolver in TEST_RESOLVERS:
     did_resolver_registry.register(resolver)
 
 resolver = DIDResolver(did_resolver_registry)
```
Comment (dbluhm) [inline aries_cloudagent/core/tests/test_did_resolver.py:84]: This should be a fixture

**Review by dbluhm — COMMENTED** [review]
Left some comments

**Comment by burdettadam** [comment]
I agree, I will move the resolver and registry into the resolver directory.

File: aries_cloudagent/resolver/did_resolver.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+LOGGER = logging.getLogger(__name__)
+
+
+class DIDResolver:
+    """did resolver singleton."""
+
+    def __init__(self, registry: DIDResolverRegistry):
+        """Initialize a `didresolver` instance."""
+        self.did_resolver_registery = registry
```
Comment (dbluhm) [inline aries_cloudagent/resolver/did_resolver.py:24]:
```suggestion
        self.did_resolver_registry = registry
```

**Pull request merged**